### PR TITLE
wires the generic scaler to kubectl

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:
the changes introduced in this commit plumbs in the generic scaler to kubectl.
note that we don't change the behaviour of kubectl.
For example it won't scale new resources. That's the end goal.
The first step is to retrofit existing code to use the generic scaler.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/scale.go [hunk 1]
```diff
@@ -53,7 +53,10 @@ type Scaler interface {
 	ScaleSimple(namespace, name string, preconditions *ScalePrecondition, newSize uint) (updatedResourceVersion string, err error)
 }
 
-func ScalerFor(kind schema.GroupKind, c internalclientset.Interface) (Scaler, error) {
+// ScalerFor gets a scaler for a given resource
+// TODO(p0lyn0mial): remove one of the case and let it fall through to GenericScaler
```
Comment (p0lyn0mial) [inline pkg/kubectl/scale.go]: before I remove one of the case I want to see the results of e2e tests.

**Comment by deads2k** [comment]
build failure or does bazel need updating?

File: pkg/kubectl/scale.go [hunk 2]
```diff
@@ -65,8 +68,12 @@ func ScalerFor(kind schema.GroupKind, c internalclientset.Interface) (Scaler, er
 		return &StatefulSetScaler{c.Apps()}, nil
 	case extensions.Kind("Deployment"), apps.Kind("Deployment"):
 		return &DeploymentScaler{c.Extensions()}, nil
+	default:
+		if scalesGetter != nil {
+			return &GenericScaler{scalesGetter, gr}, nil
+		}
 	}
-	return nil, fmt.Errorf("no scaler has been implemented for %q", kind)
+	return nil, fmt.Errorf("no scaler has been implemented or scales getter was empty for %q", kind)
```
Comment (deads2k) [inline pkg/kubectl/scale.go]: dead code

File: pkg/kubectl/scale.go (hunk 1 above)
Comment (deads2k) [reply pkg/kubectl/scale.go]: > before I remove one of the case I want to see the results of e2e tests.

Ok.  I think I'd like to see one case removed before we go through the violence of this refactor.  I understand the intermediate state, but I'd like to show progress before we push it all out.

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 3]
```diff
@@ -286,7 +287,22 @@ func (f *ring1Factory) Scaler(mapping *meta.RESTMapping) (kubectl.Scaler, error)
...
+		return nil, err
+	}
+	restClient, err := f.clientAccessFactory.RESTClient()
+	if err != nil {
+		return nil, err
+	}
+	mapper, _ := f.Object()
+	resolver := scaleclient.NewDiscoveryScaleKindResolver(discoClient)
+	scalesGetter := scaleclient.New(restClient, mapper, dynamic.LegacyAPIPathResolverFunc, resolver)
+	gvk := mapping.GroupVersionKind.GroupVersion().WithResource(mapping.Resource)
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: I suspect we want a factory method that gives back the `scalesGetter`

File: test/e2e/framework/framework.go [hunk 4]
```diff
@@ -161,6 +168,24 @@ func (f *Framework) BeforeEach() {
...
+		    config.NegotiatedSerializer = legacyscheme.Codecs
+		}
+		restClient, err := rest.RESTClientFor(config)
+		Expect(err).NotTo(HaveOccurred())
+		discoClient, err := discovery.NewDiscoveryClientForConfig(config)
+		Expect(err).NotTo(HaveOccurred())
+		cachedDiscoClient := cacheddiscovery.NewMemCacheClient(discoClient)
+		restMapper := discovery.NewDeferredDiscoveryRESTMapper(cachedDiscoClient, meta.InterfacesForUnstructured)
+		resolver := scaleclient.NewDiscoveryScaleKindResolver(cachedDiscoClient)
+		f.ScalesGetter = scaleclient.New(restClient, restMapper, dynamic.LegacyAPIPathResolverFunc,resolver)
```
Comment (deads2k) [inline test/e2e/framework/framework.go]: Looking at usage, do we really want a `scaleclient.NewDefaultForConfig(cfg *restclient.Config)` that uses "normal" defaults and builds its own discovery client and "normal" restmapper?

File: pkg/kubectl/scale.go [hunk 5]
```diff
@@ -65,8 +68,12 @@ func ScalerFor(kind schema.GroupKind, c internalclientset.Interface) (Scaler, er
 		return &StatefulSetScaler{c.Apps()}, nil
 	case extensions.Kind("Deployment"), apps.Kind("Deployment"):
 		return &DeploymentScaler{c.Extensions()}, nil
+	default:
+		if scalesGetter != nil {
```
Comment (deads2k) [inline pkg/kubectl/scale.go]: Why would we allow this to be optional?

File: test/e2e/framework/framework.go (hunk 4 above)
Comment (DirectXMan12) [reply test/e2e/framework/framework.go]: it's a pretty decent idea -- the pattern pops up a few times.  Plus, if/when we update to a properly "refreshing" discovery rest mapper, for instance, we've got fewer places to update.

File: pkg/kubectl/scale.go (hunk 5 above)
Comment (p0lyn0mial) [reply pkg/kubectl/scale.go]: agree, shouldn't be optional.

File: pkg/kubectl/cmd/util/factory_object_mapping.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/kubectl/cmd/util/factory_object_mapping.go]: alright, thanks.

**Comment by p0lyn0mial** [comment]
> build failure or does bazel need updating?

bazel :)

**Comment by p0lyn0mial** [comment]
alright, all green, I'm going to remove one case from `ScalerFor` method.

**Comment by deads2k** [comment]
> failing e2e tests look real and the primary reason being:
> error: Scaling the resource failed with: v1beta1.Scale is not suitable for converting...
> 
> What can we do about that ?

At a glance, it looks like the group isn't "autoscaling", it's empty.  I'll leave this up to remind me to look again once I get in.

**Comment by deads2k** [comment]
@p0lyn0mial just verify is failing now. Can you squash down?  I'm going to take another pass.

**Comment by deads2k** [comment]
> @p0lyn0mial just verify is failing now. Can you squash down? I'm going to take another pass.

Yeah, this all looks good enough for a first pass.  When the others are removed, we will want to ripple back through all the calls, but this is a strong start without much disruption.

/approve no-issue

**Comment by p0lyn0mial** [comment]
@deads2k many thanks for helping fixing the e2e tests. 

commits squashed, PTAL

**Comment by deads2k** [comment]
> @deads2k many thanks for helping fixing the e2e tests.
> 
> commits squashed, PTAL

This is great!  Thanks!

**Pull request merged**